### ISSUE TITLE
RPC sockets get stuck in CLOSE_WAIT state

### ISSUE THREAD

**Issue by espringe** [bitcoin/bitcoin#3968]
When RPC requests get interrupted, the sockets can end up stuck in CLOSE_WAIT state. After 3 are in this state, `bitcoind` is unable to respond to RPC requests. Happens with `0.8.6` and `0.9.0`. And lasts _far_ longer than `rpctimeout` would suggest.

```
$ lsof -p `pidof bitcoind` | grep 8332
bitcoind 872 bitcoin   23u  IPv6  10179      0t0     TCP *:8332 (LISTEN)
bitcoind 872 bitcoin   30u  IPv6  11153      0t0     TCP 107.101.212.239:8332->customer-GDL-125-2.megared.net.mx:61668 (CLOSE_WAIT)
bitcoind 872 bitcoin   38u  IPv6  11243      0t0     TCP 107.101.212.239:8332->customer-GDL-125-2.megared.net.mx:61670 (CLOSE_WAIT)
```

Googling this issue, it seems lots of people have been bitten by it -- without any real solutions

**Comment by rubensayshi** [comment]
I was experiencing similar issues, where if hammering the RPC with requests as fast as possible (from a single threaded client process) it would at some point get stuck and hit the client side timeout (60s), after a while it get's unstuck and I can do a big batch of requests again.  
 - Client process is PHP CURL (or HHVM, same result)
 - Also experienced the same issue when hammering the `0.10` rest interface.  
 - Could not see anything suspicious in the `lsof`, so I'm not sure if it's the same issue.  
 - Nothing special in the `debug.log`, with `-debug` or `debug=rpc`.  

I could not reproduce it with a simple `while [ True ]; do bitcoin-cli -testnet getrawtransaction 1030c244faabd45f6750d7c7254fa1aa87158deb900a329ef50a6dcb3aa0228e; done`, nor with `while [ True ]; do curl http://localhost:18332/rest/tx/1030c244faabd45f6750d7c7254fa1aa87158deb900a329ef50a6dcb3aa0228e.hex; done;`

I managed to fix the issue by adding a `Connection: close` header to the client CURL request, after that I can hammer it for hours without interuptions.  
Or alternatively reuse the connection handler.

@espringe what are you using to do the requests and can you reproduce the RPC getting stuck by doing an infinite loop for `getrawtransaction 1030c244faabd45f6750d7c7254fa1aa87158deb900a329ef50a6dcb3aa0228e`?  
If so, can you try adding the `Connection: close` header and see if that fixes it for you?

**Comment by sipa** [comment]
Since 0.10, rpc keepalive is optional and off by default (in 0.8 and 0.9 it was always on). Does that resolve the issue?

**Comment by mcelrath** [comment]
This is likely the same as #6454 and fixed by #5677
